### ISSUE TITLE
Semi-Structured Sparsity Causing Memory Overflow and Unsupported Operations in LLaVA Model Inference on A100 GPU

### ISSUE THREAD

**Issue by koking0** [pytorch/pytorch#117101]
## Issue description

I am following the method of this blog [Accelerating Generative AI with PyTorch: Segment Anything, Fast](https://pytorch.org/blog/accelerating-generative-ai/#sparse-semi-structured-24-sparsity) to accelerate [LLaVA](https://llava-vl.github.io/)  inference.

When I experimented with the Semi-structured sparsity solution, I encountered a problem.

First, when I apply semi-structured sparse to the entire model, the memory explodes (cuda out of memory). But my A100 has 40960 MiB of video memory, which logically shouldn't be the case.

LLaVA is composed of Llama + CLIP, and the model structure is as follows.

```
LlavaLlamaForCausalLM(
   (model): LlavaLlamaModel(
     (embed_tokens): Embedding(32000, 5120, padding_idx=0)
     (layers): ModuleList(
       (0-39): 40 x LlamaDecoderLayer(
         (self_attn): LlamaAttention(
           (q_proj): Linear(in_features=5120, out_features=5120, bias=False)
           (k_proj): Linear(in_features=5120, out_features=5120, bias=False)
           (v_proj): Linear(in_features=5120, out_features=5120, bias=False)
           (o_proj): Linear(in_features=5120, out_features=5120, bias=False)
           (rotary_emb): LlamaRotaryEmbedding()
         )
         (mlp): LlamaMLP(
           (gate_proj): Linear(in_features=5120, out_features=13824, bias=False)
           (up_proj): Linear(in_features=5120, out_features=13824, bias=False)
           (down_proj): Linear(in_features=13824, out_features=5120, bias=False)
           (act_fn): SiLUActivation()
         )
         (input_layernorm): LlamaRMSNorm()
         (post_attention_layernorm): LlamaRMSNorm()
       )
     )
     (norm): LlamaRMSNorm()
     (vision_tower): CLIPVisionTower(
       (vision_tower): CLIPVisionModel(
         (vision_model): CLIPVisionTransformer(
           (embeddings): CLIPVisionEmbeddings(
             (patch_embedding): Conv2d(3, 1024, kernel_size=(14, 14), stride=(14, 14), bias=False)
             (position_embedding): Embedding(577, 1024)
           )
           (pre_layrnorm): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)
           (encoder): CLIPEncoder(
             (layers): ModuleList(
               (0-23): 24 x CLIPEncoderLayer(
                 (self_attn): CLIPAttention(
                   (k_proj): Linear(in_features=1024, out_features=1024, bias=True)
                   (v_proj): Linear(in_features=1024, out_features=1024, bias=True)
                   (q_proj): Linear(in_features=1024, out_features=1024, bias=True)
                   (out_proj): Linear(in_features=1024, out_features=1024, bias=True)
                 )
                 (layer_norm1): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)
                 (mlp): CLIPMLP(
                   (activation_fn): QuickGELUActivation()
                   (fc1): Linear(in_features=1024, out_features=4096, bias=True)
                   (fc2): Linear(in_features=4096, out_features=1024, bias=True)
                 )
                 (layer_norm2): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)
               )
             )
           )
            (post_layernorm): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)
         )
       )
     )
     (mm_projector): Sequential(
       (0): Linear(in_features=1024, out_features=5120, bias=True)
       (1): GELU(approximate='none')
       (2): Linear(in_features=5120, out_features=5120, bias=True)
     )
   )
   (lm_head): Linear(in_features=5120, out_features=32000, bias=False)
)
```

So I performed semi-structured sparse on vision_tower alone, but encountered another error.

![image](https://github.com/pytorch/pytorch/assets/45281765/c0ee5a83-601c-4358-92ca-8b68cecc0429)

The cause of this error has not been found for the time being, so I chose to only perform semi-structured sparse on the first layer of Llama, that is, apply_sparse(model.model.layers[0]). It did not burst the video memory, but the following error was reported.

![image](https://github.com/pytorch/pytorch/assets/45281765/16b3058d-2e3c-4bd3-9644-7aa7d32ee51b)

In the Llama model structure of LLaVA, the Linear layer of the attention layer does not contain bias. It is speculated that the above error may be caused by the lack of bias.

In addition, in the PyTorch documentation, semi-structured sparse only supports the following operations, and bias is also required in Linear.

![image](https://github.com/pytorch/pytorch/assets/45281765/1ee2232c-5a8b-45b4-ac1c-0e52df8627b9)

https://pytorch.org/docs/stable/sparse.html#sparse-semi-structured-tensors

In LLaVA, except Vision Tower, only the Linear of mm_projector contains bias. Sure enough, `apply_sparse(models.model.mm_projector)` succeeded, but the inference speed did not increase.

Performing semi-structured sparse on mm_projector, memory usage: 28194 MiB, time-consuming: 149.9 ms (average of 200 pictures).
Without semi-structured sparse mm_projector, memory usage: 28034 MiB, time consumption: 150.7 ms (average of 200 pictures).

my question is:
1. Why does the video memory explode? How is semi-structured sparse converted?
2. Is it possible to make the self-attention layer in Llama support semi-structured sparse?
3. Is it possible to support semi-structured sparse in Vision Tower?

## System Info

![image](https://github.com/pytorch/pytorch/assets/45281765/2b085557-a23a-4773-9aca-92bc807b37d7)
![image](https://github.com/pytorch/pytorch/assets/45281765/195fbfb6-f2ef-45b1-8437-2ffcd02dafd1)

**Comment by bdhirsh** [comment]
cc @jcaip @cpuhrsch

**Comment by alexsamardzic** [comment]
Can you provide the exact code that you've encountered problems with?

**Comment by koking0** [comment]
> Can you provide the exact code that you've encountered problems with?

Sorry, my code is on the server and cannot be copied. I can provide a screenshot and a description.

![image](https://github.com/pytorch/pytorch/assets/45281765/b74163bc-eedf-42df-82c6-752e73474335)

Among them, from load_pretrained_model function from
https://github.com/haotian-liu/llava?tab=readme-if-quick-start-with-Huggingface

other code


```python
import torch
from torch.sparse import to_sparse_semi_structured, SparseSemiStructuredTensor
SparseSemiStructuredTensor._FORCE_CUTLASS = True

# Sparsity helper functions
def apply_fake_sparsity(model):
    """
    This function simulates 2:4 sparsity on all linear layers in a model.
    It uses the torch.ao.pruning flow.
    """
    # torch.ao.pruning flow
    from torch.ao.pruning import WeightNormSparsifier
    sparse_config = []
    for name, mod in model.named_modules():
        if isinstance(mod, torch.nn.Linear):
            sparse_config.append({"tensor_fqn": f"{name}.weight"})

    sparsifier = WeightNormSparsifier(sparsity_level=1.0,
                                      sparse_block_shape=(1,4),
                                      zeros_per_block=2)
    sparsifier.prepare(model, sparse_config)
    sparsifier.step()

    sparsifier.step()
    sparsifier.squash_mask()


def apply_sparse(model):
    apply_fake_sparsity(model)
    for name, mod in model.named_modules():
        if isinstance(mod, torch.nn.Linear):
            mod.weight = torch.nn.Parameter(to_sparse_semi_structured(mod.weight))
```

**Comment by alexsamardzic** [comment]
It's hard to reason about the issues without the code being available.

As far as use of GPU memory during the conversion to sparse semi-structured concerned:  Yes, some steps of conversion require significant amount of GPU memory (up to several times the size of the original matrix), but this memory is released as soon as conversion completed.  So if conversion performed as in `apply_sparse()` method you provided, with dense weight matrices being replaced with sparse semi-structured ones, then the GPU memory used should not be a problem, and it should be actually smaller after the conversion of all weight matrices completed.  However, if references to original weight matrices kept somewhere else, then the memory usage will indeed increase along the way, and could exceed the amount of memory available on your GPU.

As far as problem reported by `two_four_sgemm_cutlass` method, it is just probably as written in the error message, i.e. that the number of inputs provided is not divisible by 8.  You may wish to try a nightly build of PyTorch, as there were some relevant changes in the meantime.

As far as problem with `bias` argument not supplied concerned, you can just pass `None` for the bias value.

**Comment by jcaip** [comment]
Yes the nightlies have added padding support (#110744), which should fix the `two_four_sgemm_cutlass` issue. Can you try rerunning your code with them and seeing if the error goes away @koking0 ?

**Comment by alexsamardzic** [comment]
@jcaip: Shall we change the `torch.ops.aten.linear.default` operator implementation in `semi_structured.py`, so that it checks for number of arguments and set `bias` to `None` if corresponding argument not supplied?

**Comment by koking0** [comment]
To solve the problem of no bias, I set the bias to None by modifying the source code, but after the modification, the error "Number of columns of dense matrix must be divisible by 8" appeared again.

Continue to follow the reply, saying that the PyTorch nightly version can be automatically filled in. After I upgraded to torch-2.3.0.dev20240111+cu118-cp39-cp39-linux_x86_64, since Flash Attention only supports PyTorch 2.2, I cannot continue the experiment.

What is strange is this error. Each Linear layer in LLaVA, whether in_features or out_features, is a multiple of 8, but why is the error must be divisible by 8 reported?

**Comment by alexsamardzic** [comment]
Again, it's hard to tell without a reproducer...  The code first checks the divisibility (by 8 in this case), and then reports an error if the divisibility condition is not satisfied, so if an error is reported then probably the condition is not satisfied.  Would it be possible for you to find `semi_structured.py` file in your PyTorch installation, and then before each `torch._sparse_semi_structured_linear()` call, to put a `print()` call to print the input tensor shape (for example, for the first instance of this call, the print statement would be `print(input_A.shape)`, and so on)?  This way, we could see actual sizes of tensors that get checked.

**Comment by jcaip** [comment]
> @jcaip: Shall we change the torch.ops.aten.linear.default operator implementation in semi_structured.py, so that it checks for number of arguments and set bias to None if corresponding argument not supplied?

Yes, let's add a test for this at the very least, I think possible the changes I added in 2.2 might fix this but I'm not sure. 

> What is strange is this error. Each Linear layer in LLaVA, whether in_features or out_features, is a multiple of 8, but why is the error must be divisible by 8 reported?

Can you try with 2.2 then? I think that the padding changes have made it into that branch cut. To answer your question, this is because there are shape constraints on **both** the dense and sparse matrix when doing sparse @ dense matmul. In this case, it does not matter if in_features and out_features are both multiples of 8 if the dense matmul shapes are not (which is what is going on here as far as I can tell, you would expect to see a different error if the sparse matrix had the wrong shapes).

**Comment by alexsamardzic** [comment]
> > @jcaip: Shall we change the torch.ops.aten.linear.default operator implementation in semi_structured.py, so that it checks for number of arguments and set bias to None if corresponding argument not supplied?
> 
> Yes, let's add a test for this at the very least, I think it's possible the changes I added in 2.2 might fix this but I'm not sure.
> 

Indeed - calling linear operator without specifying bias seems to be working fine now.
